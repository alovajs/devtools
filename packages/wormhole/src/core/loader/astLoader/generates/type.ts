import { AST, ASTType, CommentType } from '@/type';

export interface GeneratorCtx {
  options: GeneratorOptions;
  next(ast: AST, options: GeneratorOptions): GeneratorResult;
}
export interface ASTGenerator {
  type: ASTType | ASTType[];
  generate(ast: AST, ctx: GeneratorCtx): GeneratorResult;
}
export interface GeneratorResult {
  name: string;
  type: 'type' | 'interface' | 'enum';
  comment?: string;
  code: string;
}

export interface GeneratorOptions {
  deep?: boolean;
  commentType: CommentType;
}
