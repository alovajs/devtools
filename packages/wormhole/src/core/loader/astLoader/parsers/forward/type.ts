import type { ParserSchemaType } from '../type'
import type { SchemaObject } from '@/type'

export interface Forwarder {
  is: (schema: SchemaObject) => boolean
  to: ParserSchemaType
}
