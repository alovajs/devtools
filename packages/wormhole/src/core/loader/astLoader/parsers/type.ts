import type { AST, CommentType, ReferenceObject, SchemaObject, SchemaType } from '@/type';

export type ParserSchemaType = SchemaType | 'enum' | 'group' | 'tuple' | 'reference';
export interface ASTParser {
  type: ParserSchemaType | ParserSchemaType[];
  parse(schema: SchemaObject, ctx: ParserCtx): AST;
}
export interface ParserOptions {
  commentType: CommentType;
}
export interface ParserCtx {
  next(schema: SchemaObject | ReferenceObject, options: ParserOptions): AST;
  keyName?: string;
  options: ParserOptions;
}
