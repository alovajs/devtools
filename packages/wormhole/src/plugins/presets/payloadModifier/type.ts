export type ModifierScope = 'params' | 'pathParams' | 'data' | 'response'
export type SchemaPrimitive = 'number' | 'string' | 'boolean' | 'undefined' | 'null' | 'unknown' | 'any' | 'never'

/**
 * 表示数组类型
 */
export interface SchemaArray {
  type: 'array'
  // 支持普通数组（items 为单个 schema）与元组数组（items 为 schema 数组）
  items: Schema | Schema[]
}

/**
 * 修改参数为引用类型
 * 在key末端添加上?表示为可选值
 */
export interface SchemaReference {
  [attr: string]: Schema
}

/**
 * 枚举类型表示
 */
export interface SchemaEnum {
  enum: Array<string | number | boolean | null>
  type?: 'string' | 'number' | 'integer' | 'boolean' | 'null'
}

/**
 * 组合类型表示（与/或/交叉）
 */
export interface SchemaOneOf { oneOf: Schema[] }
export interface SchemaAnyOf { anyOf: Schema[] }
export interface SchemaAllOf { allOf: Schema[] }

/**
 * 数据Schema
 * SchemaArray表示类型数组，而数组表示“或”的意思
 */
export type Schema
  = | SchemaPrimitive
    | SchemaReference
    | SchemaArray
    | SchemaEnum
    | SchemaOneOf
    | SchemaAnyOf
    | SchemaAllOf
   // 兼容旧写法：数组表示 oneOf 联合类型
    | Array<SchemaPrimitive | SchemaReference | SchemaArray | SchemaEnum>

interface ModifierConfig<T extends Schema> {
  /**
   * 生效范围，表示处理哪个位置的参数
   */
  scope: ModifierScope
  /**
   * 匹配规则，只有匹配到的才会进行转换，不指定则转换全部
   * string：原参数名包含此string；RegExp：原参数名匹配此正则；函数时接收key并返回是否匹配的boolean值
   */
  match?: string | RegExp | ((key: string) => boolean)
  /**
   * handler用于灵活修改参数类型值
   * @param schema Schema中的一种，由用户自行定义
   * @returns 返回多种参数，具体为：Schema表示修改的类型；{ required: boolean, value: Schema }表示可将当前值修改为是否必填；void | null | undefined表示移除当前字段
   */
  handler: (schema: T) => Schema | { required: boolean, value: Schema } | void | null | undefined
}

export type PayloadModifierConfig = ModifierConfig<Schema>
