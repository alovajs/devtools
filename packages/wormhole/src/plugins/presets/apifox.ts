import type { ApiPlugin } from '@/type'
import { isEmpty } from 'lodash'

export interface APIFoxBody {
  scope?: {
    type?: 'ALL' | 'SELECTED_TAGS'
    selectedTags?: string[]
    excludedByTags?: string[]
  }
  options?: {
    includeApifoxExtensionProperties?: boolean
    addFoldersToTags?: boolean
  }
  oasVersion?: '2.0' | '3.0' | '3.1'
  exportFormat?: 'JSON' | 'YAML'
  environmentIds?: string[]
}

export interface ApifoxOptions
  extends Pick<APIFoxBody, 'oasVersion' | 'exportFormat'>,
  Pick<
    NonNullable<APIFoxBody['options']>,
      'includeApifoxExtensionProperties' | 'addFoldersToTags'
  > {
  projectId: string
  apifoxToken: string
  locale?: string
  apifoxVersion?: string
  selectedTags?: string[]
  excludedByTags?: string[]
}
export function apifox({
  projectId,
  locale = 'zh-CN',
  apifoxVersion = '2024-03-28',
  selectedTags,
  excludedByTags = [],
  apifoxToken,
  oasVersion = '3.0',
  exportFormat = 'JSON',
  includeApifoxExtensionProperties = false,
  addFoldersToTags = false,
}: ApifoxOptions): ApiPlugin {
  const body: APIFoxBody = {
    scope: {
      excludedByTags,
    },
    options: {
      includeApifoxExtensionProperties,
      addFoldersToTags,
    },
    oasVersion,
    exportFormat,
  }
  if (!body.scope) {
    body.scope = {}
  }
  const tags = !isEmpty(selectedTags) ? selectedTags : '*'
  if (tags === '*') {
    body.scope.type = 'ALL'
  }
  else {
    body.scope.type = 'SELECTED_TAGS'
    body.scope.selectedTags = tags
  }
  return {
    name: 'apifox',
    config(config) {
      const base = 'https://api.apifox.com/v1/projects'
      if (projectId && apifoxToken) {
        config.input = `${base}/${encodeURIComponent(projectId)}/export-openapi?locale=${encodeURIComponent(locale)}`
        config.fetchOptions = {
          ...config.fetchOptions,
          headers: {
            'X-Apifox-Api-Version': apifoxVersion,
            'Authorization': `Bearer ${apifoxToken}`,
          },
          method: 'POST',
          data: body,
        }
      }
      return config
    },
  }
}

export default apifox
