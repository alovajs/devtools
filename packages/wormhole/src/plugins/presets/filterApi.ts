import { logger } from '@/helper/logger';
import type { ApiDescriptor, ApiPlugin } from '@/type';
/**
 * Filter configuration interface
 */
export interface FilterApiConfig {
  /**
   * Target scope for filtering, defaults to 'url'
   */
  scope?: 'url' | 'tag';

  /**
   * Include rule:
   * - string: target contains this string
   * - RegExp: target matches this pattern
   * - function: custom matching logic
   */
  include?: string | RegExp | ((key: string) => boolean);

  /**
   * Exclude rule:
   * - string: target contains this string
   * - RegExp: target matches this pattern
   * - function: custom matching logic
   */
  exclude?: string | RegExp | ((key: string) => boolean);
}

/**
 * Tests if value matches the specified rule
 */
function isMatch(value: string, match?: string | RegExp | ((key: string) => boolean)): boolean {
  if (!match) return true;

  if (typeof match === 'string') {
    return value.includes(match);
  }

  if (match instanceof RegExp) {
    return match.test(value);
  }

  if (typeof match === 'function') {
    try {
      return match(value);
    } catch {
      return false; // Return false on error to exclude the item
    }
  }

  return false;
}

/**
 * Extracts the corresponding property value from API descriptor based on scope
 */
function getApiProperty(apiDescriptor: ApiDescriptor, scope: 'url' | 'tag'): string {
  switch (scope) {
    case 'url':
      return apiDescriptor.url || '';
    case 'tag':
      // Assume tags exist in tags array, join with comma if multiple tags
      return Array.isArray(apiDescriptor.tags) ? apiDescriptor.tags.join(',') : '';
    default:
      return '';
  }
}

/**
 * Applies filtering rules for a single configuration
 * @param apiDescriptor API descriptor
 * @param config Filter configuration
 * @returns Whether it passes the filter (true means keep, false means filter out)
 */
function applyFilterRule(apiDescriptor: ApiDescriptor, config: FilterApiConfig): boolean {
  const scope = config.scope || 'url';
  const value = getApiProperty(apiDescriptor, scope);

  // Handle include and exclude logic
  const includeMatch = config.include ? isMatch(value, config.include) : true;
  const excludeMatch = config.exclude ? isMatch(value, config.exclude) : false;

  // If both include and exclude are specified, exclude matching items from include
  return includeMatch && !excludeMatch;
}

/**
 * Handles union logic for multiple configurations
 * @param apiDescriptor API descriptor
 * @param configs Configuration array
 * @returns Whether it passes the filter (true means keep, false means filter out)
 */
function combineFilterResults(apiDescriptor: ApiDescriptor, configs: FilterApiConfig[]): boolean {
  // If any configuration matches, keep the API (union logic)
  return configs.some(config => applyFilterRule(apiDescriptor, config));
}

/**
 * Main processing function for filtering API descriptors
 * @param apiDescriptor API descriptor
 * @param configs Configuration array
 * @returns Filtered API descriptor, or null if filtered out
 */
export function filterApiDescriptor(apiDescriptor: ApiDescriptor, configs: FilterApiConfig[]): ApiDescriptor | null {
  if (!apiDescriptor) return null;

  // Use union logic to determine whether to keep the API
  const shouldKeep = combineFilterResults(apiDescriptor, configs);

  return shouldKeep ? apiDescriptor : null;
}

/**
 * Creates a plugin for filtering APIs
 *
 * @param config Filter configuration, can be a single config or array of configs
 * @returns API plugin instance
 *
 * @example
 * ```ts
 * // Only include URLs containing 'user'
 * const userOnlyFilter = apiFilter({
 *   include: 'user'
 * });
 *
 * // Exclude tags containing 'internal'
 * const noInternalFilter = apiFilter({
 *   scope: 'tag',
 *   exclude: 'internal'
 * });
 *
 * // Multi-condition filtering (union)
 * const multiFilter = apiFilter([
 *   { include: 'user' },
 *   { include: 'admin' }
 * ]);
 * ```
 */
export function apiFilter(config: FilterApiConfig | FilterApiConfig[]): ApiPlugin {
  const configs = Array.isArray(config) ? config : [config];

  // Validate configuration
  for (const conf of configs) {
    if (!conf.include && !conf.exclude) {
      throw logger.throwError('at least one of `include` or `exclude` must be specified');
    }
  }

  return {
    name: 'filterApi',
    extends: {
      handleApi: (apiDescriptor: ApiDescriptor) => {
        if (!apiDescriptor) return null;

        return filterApiDescriptor(apiDescriptor, configs);
      }
    }
  };
}

export default apiFilter;
