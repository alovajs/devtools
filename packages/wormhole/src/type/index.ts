export * from './api';
export * from './base';
export * from './openapi';
