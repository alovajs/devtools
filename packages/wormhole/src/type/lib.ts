export { Api, ApiDescriptor, ApiDoc, TemplateData } from './api'
export * from '@/helper/config/type'
export interface GenerateApiOptions {
  force?: boolean
  projectPath?: string
}
