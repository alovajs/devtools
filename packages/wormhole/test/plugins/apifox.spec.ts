import type { GeneratorConfig } from '@/type'
import { describe, expect, it } from 'vitest'
import { apifox } from '@/plugins/presets/apifox'
import { generateWithPlugin } from '../util'

vi.mock('node:fs')
vi.mock('node:fs/promises')
describe('apifox preset plugin - beforeOpenapiParse', () => {
  const baseInputConfig: Pick<GeneratorConfig, 'input' | 'platform' | 'plugins' | 'fetchOptions'> = {
    input: 'dummy',
    platform: 'swagger',
    plugins: [],
  }

  it('should set ALL scope when selectedTags not provided and construct request', async () => {
    const plugin = apifox({
      projectId: 'proj-123',
      apifoxToken: 'token-abc',
    })

    const next = (await plugin.beforeOpenapiParse!(baseInputConfig)) ?? baseInputConfig

    // input URL
    expect(next.input).toBe(
      'https://api.apifox.com/v1/projects/proj-123/export-openapi?locale=zh-CN',
    )

    // headers
    expect(next.fetchOptions?.headers).toEqual({
      'X-Apifox-Api-Version': '2024-03-28',
      'Authorization': 'Bearer token-abc',
    })

    // body defaults
    expect(next.fetchOptions?.data).toEqual({
      scope: {
        excludedByTags: [],
        type: 'ALL',
      },
      options: {
        includeApifoxExtensionProperties: false,
        addFoldersToTags: false,
      },
      oasVersion: '3.0',
      exportFormat: 'JSON',
    })
  })

  it('should set SELECTED_TAGS scope when selectedTags provided', async () => {
    const plugin = apifox({
      projectId: 'proj-123',
      apifoxToken: 'token-abc',
      selectedTags: ['user', 'order'],
    })

    const next = (await plugin.beforeOpenapiParse!(baseInputConfig)) ?? baseInputConfig
    const data = next.fetchOptions?.data as Record<string, any>
    expect(data.scope?.type).toBe('SELECTED_TAGS')
    expect(data.scope?.selectedTags).toEqual(['user', 'order'])
  })

  it('should encode locale and projectId, and accept overrides', async () => {
    const plugin = apifox({
      projectId: '中文 项目',
      apifoxToken: 'token-abc',
      locale: 'en-US',
      apifoxVersion: '2025-01-01',
      includeApifoxExtensionProperties: true,
      addFoldersToTags: true,
      oasVersion: '3.1',
      exportFormat: 'YAML',
      excludedByTags: ['internal'],
    })

    const next = (await plugin.beforeOpenapiParse!(baseInputConfig)) ?? baseInputConfig

    expect(next.input).toBe(
      'https://api.apifox.com/v1/projects/%E4%B8%AD%E6%96%87%20%E9%A1%B9%E7%9B%AE/export-openapi?locale=en-US',
    )
    expect(next.fetchOptions?.headers).toEqual({
      'X-Apifox-Api-Version': '2025-01-01',
      'Authorization': 'Bearer token-abc',
    })
    expect(next.fetchOptions?.data).toEqual({
      scope: {
        excludedByTags: ['internal'],
        type: 'ALL',
      },
      options: {
        includeApifoxExtensionProperties: true,
        addFoldersToTags: true,
      },
      oasVersion: '3.1',
      exportFormat: 'YAML',
    })
  })

  it('should not modify config when missing projectId or token', async () => {
    const pluginMissingProject = apifox({ apifoxToken: 't', projectId: '' })
    const pluginMissingToken = apifox({ projectId: 'p', apifoxToken: '' })

    const next1 = (await pluginMissingProject.beforeOpenapiParse!(baseInputConfig)) ?? baseInputConfig
    const next2 = (await pluginMissingToken.beforeOpenapiParse!(baseInputConfig)) ?? baseInputConfig

    expect(next1).toEqual(baseInputConfig)
    expect(next2).toEqual(baseInputConfig)
    expect(next1.platform).toBe('swagger')
    expect(next2.plugins).toEqual([])
  })

  it('should expose plugin name', () => {
    const plugin = apifox({ projectId: 'p', apifoxToken: 't' })
    expect(plugin.name).toBe('apifox')
  })

  // Integration: actually use apifox via generateWithPlugin and MSW
  it('should generate using Apifox export endpoint via MSW', async () => {
    const { apiDefinitionsFile } = await generateWithPlugin(
      'test-apifox',
      [
        apifox({
          projectId: 'proj-123',
          apifoxToken: 'token-abc',
        }),
      ],
    )
    // The generated definitions should reflect Apifox OpenAPI content
    expect(apiDefinitionsFile).toMatchSnapshot()
  })
})
