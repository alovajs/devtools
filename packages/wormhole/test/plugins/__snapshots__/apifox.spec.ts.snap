// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`apifox preset plugin - beforeOpenapiParse > should generate using Apifox export endpoint via MSW 1`] = `
"/// <reference types='./globals.d.ts' />
/* tslint:disable */
/* eslint-disable */
/**
 * Swagger Generator - version 3.0.57
 *
 * This is an online swagger codegen server.  You can find out more at https://github.com/swagger-api/swagger-codegen or on [irc.freenode.net, #swagger](http://swagger.io/irc/).
 *
 * OpenAPI version: 3.0.1
 *
 *
 * NOTE: This file is auto generated by the alova's vscode plugin.
 *
 * https://alova.js.org/devtools/vscode
 *
 * **Do not edit the file manually.**
 */
export default {
  'clients.generateFromURL': ['GET', '/generate'],
  'servers.generateFromURL': ['GET', '/generate'],
  'documentation.generateFromURL': ['GET', '/generate'],
  'config.generateFromURL': ['GET', '/generate'],
  'clients.generate': ['POST', '/generate'],
  'servers.generate': ['POST', '/generate'],
  'documentation.generate': ['POST', '/generate'],
  'config.generate': ['POST', '/generate'],
  'clients.clientLanguages': ['GET', '/clients'],
  'documentation.clientLanguages': ['GET', '/clients'],
  'servers.serverLanguages': ['GET', '/servers'],
  'documentation.documentationLanguages': ['GET', '/documentation'],
  'clients.languages': ['GET', '/{type}/{version}'],
  'servers.languages': ['GET', '/{type}/{version}'],
  'documentation.languages': ['GET', '/{type}/{version}'],
  'config.languages': ['GET', '/{type}/{version}'],
  'clients.languagesMulti': ['GET', '/types'],
  'servers.languagesMulti': ['GET', '/types'],
  'documentation.languagesMulti': ['GET', '/types'],
  'config.languagesMulti': ['GET', '/types'],
  'clients.listOptions': ['GET', '/options'],
  'servers.listOptions': ['GET', '/options'],
  'documentation.listOptions': ['GET', '/options'],
  'config.listOptions': ['GET', '/options'],
  'clients.generateBundle': ['POST', '/model'],
  'servers.generateBundle': ['POST', '/model'],
  'documentation.generateBundle': ['POST', '/model'],
  'config.generateBundle': ['POST', '/model'],
  'documentation.renderTemplate': ['POST', '/render']
};
"
`;

exports[`apifox preset plugin - config > should generate using Apifox export endpoint via MSW 1`] = `
"/// <reference types='./globals.d.ts' />
/* tslint:disable */
/* eslint-disable */
/**
 * Swagger Generator - version 3.0.57
 *
 * This is an online swagger codegen server.  You can find out more at https://github.com/swagger-api/swagger-codegen or on [irc.freenode.net, #swagger](http://swagger.io/irc/).
 *
 * OpenAPI version: 3.0.1
 *
 *
 * NOTE: This file is auto generated by the alova's vscode plugin.
 *
 * https://alova.js.org/devtools/vscode
 *
 * **Do not edit the file manually.**
 */
export default {
  'clients.generateFromURL': ['GET', '/generate'],
  'servers.generateFromURL': ['GET', '/generate'],
  'documentation.generateFromURL': ['GET', '/generate'],
  'config.generateFromURL': ['GET', '/generate'],
  'clients.generate': ['POST', '/generate'],
  'servers.generate': ['POST', '/generate'],
  'documentation.generate': ['POST', '/generate'],
  'config.generate': ['POST', '/generate'],
  'clients.clientLanguages': ['GET', '/clients'],
  'documentation.clientLanguages': ['GET', '/clients'],
  'servers.serverLanguages': ['GET', '/servers'],
  'documentation.documentationLanguages': ['GET', '/documentation'],
  'clients.languages': ['GET', '/{type}/{version}'],
  'servers.languages': ['GET', '/{type}/{version}'],
  'documentation.languages': ['GET', '/{type}/{version}'],
  'config.languages': ['GET', '/{type}/{version}'],
  'clients.languagesMulti': ['GET', '/types'],
  'servers.languagesMulti': ['GET', '/types'],
  'documentation.languagesMulti': ['GET', '/types'],
  'config.languagesMulti': ['GET', '/types'],
  'clients.listOptions': ['GET', '/options'],
  'servers.listOptions': ['GET', '/options'],
  'documentation.listOptions': ['GET', '/options'],
  'config.listOptions': ['GET', '/options'],
  'clients.generateBundle': ['POST', '/model'],
  'servers.generateBundle': ['POST', '/model'],
  'documentation.generateBundle': ['POST', '/model'],
  'config.generateBundle': ['POST', '/model'],
  'documentation.renderTemplate': ['POST', '/render']
};
"
`;
